File: code/Java/interpreter/expr/Variable.java
package interpreter.expr;

import java.util.HashMap;
import java.util.Map;

public class Variable extends IntExpr {
    
    public static Map<String, Variable> variables;

    static {
        variables = new HashMap<String, Variable>();
    }

    private String name;
    private int value;

    private Variable(String name) {
        super(-1);

        this.name = name;
        this.value = 0;
    }

    public String getName() {
        return name;
    }

    public int expr() {
        return value;
    }

    public void setValue(int value) {
        this.value = value;
    }

    public static Variable instance(String name) {
        Variable v = variables.get(name);
        if (v == null) {
            v = new Variable(name);
            variables.put(name, v);
        }

        return v;
    }

}

File: code/C++/interpreter/expr/Variable.cpp
#include "Variable.h"
#include "../util/Memory.h"

Variable::Variable(int line, const std::string& name)
	: IntExpr(line), m_name(name) {
}

Variable::~Variable() {
}

int Variable::value() {
	return Memory::read(m_name);
}

void Variable::setValue(int value) {
	Memory::write(m_name, value);
}

int Variable::expr() {
	return this->value();
}
